core/arch/arm/mm/core_mmu.c

```diff
@@ -368,16 +368,22 @@ void core_mmu_set_discovered_nsec_ddr(struct core_mmu_phys_mem *start,
 	 * are used for a specific purpose make holes for
 	 * those memory in the normal non-secure memory.
 	 *
+	 * When device tree is used, the dtb file is
+	 * mapped by secureos as IO_NSEC memory type,
+	 * which may resides in normal non-secure memory
+	 * and thus need to be removed.
+	 *
 	 * This has to be done since for instance QEMU
 	 * isn't aware of which memory range in the
 	 * non-secure memory is used for NSEC_SHM.
 	 */
-
 #ifdef CFG_SECURE_DATA_PATH
 	for (pmem = phys_sdp_mem_begin; pmem < phys_sdp_mem_end; pmem++)
 		carve_out_phys_mem(&m, &num_elems, pmem->addr, pmem->size);
 #endif
-
+#ifdef CFG_DT
+	carve_out_phys_mem(&m, &num_elems, dt_addr, CFG_DTB_MAX_SIZE);
```

Comment:
This will make it impossible for normal world to reuse this memory as memory shared with secure world. We probably need a special memory type for memory mapped for ns-dt instead which we then can ignore in the loop below.

Reply:
@jenswi-linaro , you're right. Thanks for your advice. I'll submit another patch with your consideration.

Reply:
@jenswi-linaro , I've submitted a new patch. Please help review it. thanks. BTW, I'm new to qemu, could you tell me where's the secureos' external dts file for qemu boot?